helix-view/src/editor.rs

```diff
@@ -771,6 +775,112 @@ impl WhitespaceRender {
     }
 }
 
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct AutoSave {
+    /// Auto save after `n` milli seconds. Defaults to None.
+    pub after_delay: Option<u64>,
+    /// Auto save on focus lost. Defaults to false.
+    pub focus_lost: bool,
+}
+
+// Intermediate type to have a flattened config
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
+#[serde(untagged, deny_unknown_fields, rename_all = "kebab-case")]
+enum AutoSaveToml {
+    EnableFocusLost(bool),
+    #[serde(rename_all = "kebab-case")]
+    AutoSave {
+        #[serde(default)]
+        after_delay: AutoSaveAfterDelay,
+        #[serde(default)]
+        focus_lost: bool,
+    },
+}
+
+// Intermediate type to have a flattened config
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
+#[serde(deny_unknown_fields)]
+struct AutoSaveAfterDelay {
+    #[serde(default)]
+    pub enable: bool,
+    #[serde(default = "default_auto_save_delay")]
+    pub timeout: u64,
+}
+
+fn default_auto_save_delay() -> u64 {
+    DEFAULT_AUTO_SAVE_DELAY
+}
+
+impl From<AutoSave> for AutoSaveToml {
+    fn from(value: AutoSave) -> Self {
+        let after_delay = match value.after_delay {
+            Some(timeout) => AutoSaveAfterDelay {
+                enable: true,
+                timeout,
+            },
+            None => Default::default(),
+        };
+        Self::AutoSave {
+            after_delay,
+            focus_lost: value.focus_lost,
+        }
+    }
+}
+
+impl From<AutoSaveToml> for AutoSave {
+    fn from(value: AutoSaveToml) -> Self {
+        match value {
+            AutoSaveToml::EnableFocusLost(focus_lost) => Self {
+                focus_lost,
+                ..Default::default()
+            },
+            AutoSaveToml::AutoSave {
+                after_delay,
+                focus_lost,
+            } => {
+                let after_delay: Option<u64> = after_delay.enable.then_some(after_delay.timeout);
+                Self {
+                    after_delay,
+                    focus_lost,
+                }
+            }
+        }
+    }
+}
+
+impl Default for AutoSaveToml {
+    fn default() -> Self {
+        Self::EnableFocusLost(false)
+    }
+}
+
+impl Default for AutoSaveAfterDelay {
+    fn default() -> Self {
+        Self {
+            enable: false,
+            timeout: DEFAULT_AUTO_SAVE_DELAY,
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for AutoSave {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Ok(AutoSaveToml::deserialize(deserializer)?.into())
+    }
+}
+impl Serialize for AutoSave {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let a: AutoSaveToml = self.clone().into();
+        a.serialize(serializer)
+    }
+}
```

Comment:
This is a bit much boiler plate for my liking. I usually just put the `enum AutoSaveToml {` inside a standalone deserialize function. You can just but a `#[serde(deserialize_with = "deserialize_auto_save")]` attribute on the `auto_save` field. Combined with the other changes I roughly had the following in mind:


```suggestion
fn deserialize_auto_save<'de, D>(deserializer: D) -> Result<AutoSavei, D::Error>
where
    D: serde::Deserializer<'de>,
{
	#[derive(Deserialize, Serialize)]
	#[serde(untagged, deny_unknown_fields, rename_all = "kebab-case")]
	enum AutoSaveToml {
	    EnableFocusLost(bool),
	    AutoSave(AutoSave),
	}
	
    match AutoSaveToml::deserilaize(deserializer)? {
	    AutoSaveToml::EnableFocusLost(focus_lost) => Ok(AutoSave {
	        focus_lost,
	        ..Default::default()
	    }),
	    AutoSaveToml::AutoSave(auto_save) => Ok(auto_save)
    }
}

```

We actually want to just use the derived `Serialize` implementation for AutoSave instead of going back to `AutoSaveToml` so you don't need to do this for serialize

Reply:
The custom Default impl is only needed for AutoSaveAfterDelay in order to not have the timeout default to 0.
When setting `auto-save.after-delay.enable` to true at runtime `timeout` would be at 0 and not at the default.
Same goes for when just doing `:get-option auto-save.after-delay.timeout`.

Reply:
yeah the Default impl for `AutoSaveAfterDelay` makes sense I just coarsly selected the whole area since my comments pertained to most impls there.